crates/core/src/contract/handler.rs

```diff
@@ -273,12 +275,22 @@ impl ContractHandlerChannel<WaitingResolution> {
     pub async fn relay_transaction_result_to_client(
         &mut self,
     ) -> anyhow::Result<(ClientId, WaitingTransaction)> {
-        self.end
+        let rx = self
+            .end
             .wait_for_res_rx
-            .recv()
+            .as_mut()
+            .ok_or_else(|| anyhow::anyhow!("receiver already taken"))?;
+        rx.recv()
             .await
             .ok_or_else(|| anyhow::anyhow!("channel dropped"))
     }
+
+    /// Take the underlying receiver for streaming. Returns None if already taken.
+    pub fn take_wait_for_res_receiver(
```

Comment:
why do we need to do this? are we doing this concurrently? would it may more sense moving from mpsc to a broadcast?